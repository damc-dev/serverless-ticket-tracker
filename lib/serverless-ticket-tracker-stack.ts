import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as apigateway from '@aws-cdk/aws-apigateway'
import * as path from 'path';
import * as dynamodb from '@aws-cdk/aws-dynamodb';
import { CfnOutput, Duration } from '@aws-cdk/core';
import * as codepipeline from '@aws-cdk/aws-codepipeline'

export class ServerlessTicketTrackerStack extends cdk.Stack {
  public readonly urlOutput: CfnOutput;

  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const table = new dynamodb.Table(this, 'TicketsTable', {
      tableName: 'SttTickets',
      partitionKey: { name: 'id', type: dynamodb.AttributeType.STRING },
    });

    const api = new apigateway.RestApi(this, 'TicketsApi');
    const ticket = api.root.addResource('ticket');
    
    const getTicketHandler = new lambda.Function(this, 'GetTicket', {
      runtime: lambda.Runtime.NODEJS_10_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset('lambda/tickets-get'),
      environment: {
        TABLE_NAME: table.tableName
      },
      timeout: Duration.seconds(30)
    });
    table.grantReadData(getTicketHandler);
    ticket.addMethod('GET', new apigateway.LambdaIntegration(getTicketHandler))

    const postTicketHandler = new lambda.Function(this, 'TicketsPost', {
      runtime: lambda.Runtime.NODEJS_10_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset('lambda/tickets-post'),
      environment: {
        TABLE_NAME: table.tableName
      },
      timeout: Duration.seconds(30)
    });
    table.grantWriteData(postTicketHandler);
    ticket.addMethod('POST', new apigateway.LambdaIntegration(postTicketHandler))

    const health = api.root.addResource('health');
    const healthCheckHandler = new lambda.Function(this, 'HealthCheck', {
      runtime: lambda.Runtime.NODEJS_10_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset('lambda/health-check'),
      environment: {
        TABLE_NAME: table.tableName
      },
      timeout: Duration.seconds(30)
    });
    table.grantReadData(healthCheckHandler);
    health.addMethod('GET', new apigateway.LambdaIntegration(healthCheckHandler))

    this.urlOutput = new CfnOutput(this, 'Url', {
      value: api.url
    });
  }
}