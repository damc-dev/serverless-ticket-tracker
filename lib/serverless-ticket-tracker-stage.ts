
import { CfnOutput, Stage, Construct, StageProps } from '@aws-cdk/core';
import { ServerlessTicketTrackerStack } from './serverless-ticket-tracker-stack';

export class ServerlessTicketTrackerStage extends Stage {
  public readonly urlOutput: CfnOutput;

  constructor(scope: Construct, id: string, props?: StageProps) {
    super(scope, id, props);

    const service = new ServerlessTicketTrackerStack(this, 'ServerlessTicketTracker', {
      env: { account: process.env.CDK_DEFAULT_ACCOUNT, region: process.env.CDK_DEFAULT_REGION },
    });
    
    this.urlOutput = service.urlOutput;
  }
}