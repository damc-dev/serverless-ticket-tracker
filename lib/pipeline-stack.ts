import * as cdk from '@aws-cdk/core';
import { StackProps, Construct, SecretValue } from '@aws-cdk/core';
import * as codepipeline from '@aws-cdk/aws-codepipeline';
import * as codepipeline_actions from '@aws-cdk/aws-codepipeline-actions';
import { CdkPipeline, SimpleSynthAction } from '@aws-cdk/pipelines'
import { ServerlessTicketTrackerStack } from './serverless-ticket-tracker-stack';
import { ServerlessTicketTrackerStage } from './serverless-ticket-tracker-stage';

export class PipelineStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);
        const sourceArtifact = new codepipeline.Artifact();
        const cloudAssemblyArtifact = new codepipeline.Artifact();

        const pipeline = new CdkPipeline(this, 'Pipeline', {
          pipelineName: 'ServerlessTicketTrackerPipeline',
          cloudAssemblyArtifact,
    
          sourceAction: new codepipeline_actions.GitHubSourceAction({
            actionName: 'GitHub',
            output: sourceArtifact,
            oauthToken: SecretValue.secretsManager('github-token'),
            trigger: codepipeline_actions.GitHubTrigger.POLL,
            owner: 'damc-dev',
            repo: 'serverless-ticket-tracker',
          }),
    
          synthAction: SimpleSynthAction.standardYarnSynth({
            sourceArtifact,
            cloudAssemblyArtifact,
            buildCommand: 'yarn build',
          }),
        });
        pipeline.addApplicationStage(new ServerlessTicketTrackerStage(this, 'PreProd', {
            env: { account: '342431730723', region: 'us-east-1' }
          }));
    }
}