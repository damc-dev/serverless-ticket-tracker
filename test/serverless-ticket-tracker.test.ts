import { SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as ServerlessTicketTracker from '../lib/serverless-ticket-tracker-stack';

test('Stack Snapshot', () => {
    const app = new cdk.App();
    // WHEN
    const stack = new ServerlessTicketTracker.ServerlessTicketTrackerStack(app, 'MyTestStack');
    // THEN
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
