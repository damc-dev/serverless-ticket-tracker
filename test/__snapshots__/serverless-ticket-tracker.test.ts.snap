// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack Snapshot 1`] = `
Object {
  "Outputs": Object {
    "TicketsApiEndpoint73949A61": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "TicketsApi5BB0C274",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "TicketsApiDeploymentStageprod5BCA3CD4",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters6c04035e79eda3529ba6067e0c9de878c5d1ba171c28aae3e1639c1cd64de0dcArtifactHashAB8DD616": Object {
      "Description": "Artifact hash for asset \\"6c04035e79eda3529ba6067e0c9de878c5d1ba171c28aae3e1639c1cd64de0dc\\"",
      "Type": "String",
    },
    "AssetParameters6c04035e79eda3529ba6067e0c9de878c5d1ba171c28aae3e1639c1cd64de0dcS3Bucket76818549": Object {
      "Description": "S3 bucket for asset \\"6c04035e79eda3529ba6067e0c9de878c5d1ba171c28aae3e1639c1cd64de0dc\\"",
      "Type": "String",
    },
    "AssetParameters6c04035e79eda3529ba6067e0c9de878c5d1ba171c28aae3e1639c1cd64de0dcS3VersionKey46A45481": Object {
      "Description": "S3 key for asset version \\"6c04035e79eda3529ba6067e0c9de878c5d1ba171c28aae3e1639c1cd64de0dc\\"",
      "Type": "String",
    },
    "AssetParametersaefdea884930f8e31c77b692a0c9d29dc35671165502f5eb7303eae0df30f0f3ArtifactHash9F96D9AD": Object {
      "Description": "Artifact hash for asset \\"aefdea884930f8e31c77b692a0c9d29dc35671165502f5eb7303eae0df30f0f3\\"",
      "Type": "String",
    },
    "AssetParametersaefdea884930f8e31c77b692a0c9d29dc35671165502f5eb7303eae0df30f0f3S3BucketB7C90BB0": Object {
      "Description": "S3 bucket for asset \\"aefdea884930f8e31c77b692a0c9d29dc35671165502f5eb7303eae0df30f0f3\\"",
      "Type": "String",
    },
    "AssetParametersaefdea884930f8e31c77b692a0c9d29dc35671165502f5eb7303eae0df30f0f3S3VersionKey8F500E6E": Object {
      "Description": "S3 key for asset version \\"aefdea884930f8e31c77b692a0c9d29dc35671165502f5eb7303eae0df30f0f3\\"",
      "Type": "String",
    },
    "AssetParametersd399a49e9a3b5bf091059e38e7b6d307b3346e6e9af3896bdc1c80e33cff6e4cArtifactHashD74DFAAA": Object {
      "Description": "Artifact hash for asset \\"d399a49e9a3b5bf091059e38e7b6d307b3346e6e9af3896bdc1c80e33cff6e4c\\"",
      "Type": "String",
    },
    "AssetParametersd399a49e9a3b5bf091059e38e7b6d307b3346e6e9af3896bdc1c80e33cff6e4cS3BucketB891A3BB": Object {
      "Description": "S3 bucket for asset \\"d399a49e9a3b5bf091059e38e7b6d307b3346e6e9af3896bdc1c80e33cff6e4c\\"",
      "Type": "String",
    },
    "AssetParametersd399a49e9a3b5bf091059e38e7b6d307b3346e6e9af3896bdc1c80e33cff6e4cS3VersionKey2FC59AFE": Object {
      "Description": "S3 key for asset version \\"d399a49e9a3b5bf091059e38e7b6d307b3346e6e9af3896bdc1c80e33cff6e4c\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "GetTicket4BEE878E": Object {
      "DependsOn": Array [
        "GetTicketServiceRoleDefaultPolicy3B14A848",
        "GetTicketServiceRole3B0B5142",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersd399a49e9a3b5bf091059e38e7b6d307b3346e6e9af3896bdc1c80e33cff6e4cS3BucketB891A3BB",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersd399a49e9a3b5bf091059e38e7b6d307b3346e6e9af3896bdc1c80e33cff6e4cS3VersionKey2FC59AFE",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersd399a49e9a3b5bf091059e38e7b6d307b3346e6e9af3896bdc1c80e33cff6e4cS3VersionKey2FC59AFE",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "TABLE_NAME": Object {
              "Ref": "TicketsTableB76A19AF",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "GetTicketServiceRole3B0B5142",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "GetTicketServiceRole3B0B5142": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GetTicketServiceRoleDefaultPolicy3B14A848": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "TicketsTableB76A19AF",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GetTicketServiceRoleDefaultPolicy3B14A848",
        "Roles": Array [
          Object {
            "Ref": "GetTicketServiceRole3B0B5142",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "HealthCheckA1C381C7": Object {
      "DependsOn": Array [
        "HealthCheckServiceRoleDefaultPolicyAE8FE253",
        "HealthCheckServiceRoleC964F23C",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters6c04035e79eda3529ba6067e0c9de878c5d1ba171c28aae3e1639c1cd64de0dcS3Bucket76818549",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters6c04035e79eda3529ba6067e0c9de878c5d1ba171c28aae3e1639c1cd64de0dcS3VersionKey46A45481",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters6c04035e79eda3529ba6067e0c9de878c5d1ba171c28aae3e1639c1cd64de0dcS3VersionKey46A45481",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "TABLE_NAME": Object {
              "Ref": "TicketsTableB76A19AF",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "HealthCheckServiceRoleC964F23C",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "HealthCheckServiceRoleC964F23C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "HealthCheckServiceRoleDefaultPolicyAE8FE253": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "TicketsTableB76A19AF",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "HealthCheckServiceRoleDefaultPolicyAE8FE253",
        "Roles": Array [
          Object {
            "Ref": "HealthCheckServiceRoleC964F23C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TicketsApi5BB0C274": Object {
      "Properties": Object {
        "Name": "TicketsApi",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "TicketsApiAccount2E13CE06": Object {
      "DependsOn": Array [
        "TicketsApi5BB0C274",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "TicketsApiCloudWatchRoleCD431BD7",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "TicketsApiCloudWatchRoleCD431BD7": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TicketsApiDeployment7BF22014e1267a19b37c38a5f8ca9aaac786d7c9": Object {
      "DependsOn": Array [
        "TicketsApihealthGETB0F582E9",
        "TicketsApihealth1A62B535",
        "TicketsApiticketGET65DC6613",
        "TicketsApiticketPOST4B2578D0",
        "TicketsApiticketA33D88A0",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "TicketsApi5BB0C274",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "TicketsApiDeploymentStageprod5BCA3CD4": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "TicketsApiDeployment7BF22014e1267a19b37c38a5f8ca9aaac786d7c9",
        },
        "RestApiId": Object {
          "Ref": "TicketsApi5BB0C274",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "TicketsApihealth1A62B535": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "TicketsApi5BB0C274",
            "RootResourceId",
          ],
        },
        "PathPart": "health",
        "RestApiId": Object {
          "Ref": "TicketsApi5BB0C274",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "TicketsApihealthGETApiPermissionMyTestStackTicketsApi42B25C87GEThealth61AD6E31": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "HealthCheckA1C381C7",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "TicketsApi5BB0C274",
              },
              "/",
              Object {
                "Ref": "TicketsApiDeploymentStageprod5BCA3CD4",
              },
              "/GET/health",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TicketsApihealthGETApiPermissionTestMyTestStackTicketsApi42B25C87GEThealth369618F9": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "HealthCheckA1C381C7",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "TicketsApi5BB0C274",
              },
              "/test-invoke-stage/GET/health",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TicketsApihealthGETB0F582E9": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "HealthCheckA1C381C7",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "TicketsApihealth1A62B535",
        },
        "RestApiId": Object {
          "Ref": "TicketsApi5BB0C274",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "TicketsApiticketA33D88A0": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "TicketsApi5BB0C274",
            "RootResourceId",
          ],
        },
        "PathPart": "ticket",
        "RestApiId": Object {
          "Ref": "TicketsApi5BB0C274",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "TicketsApiticketGET65DC6613": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "GetTicket4BEE878E",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "TicketsApiticketA33D88A0",
        },
        "RestApiId": Object {
          "Ref": "TicketsApi5BB0C274",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "TicketsApiticketGETApiPermissionMyTestStackTicketsApi42B25C87GETticketE19F28FB": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "GetTicket4BEE878E",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "TicketsApi5BB0C274",
              },
              "/",
              Object {
                "Ref": "TicketsApiDeploymentStageprod5BCA3CD4",
              },
              "/GET/ticket",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TicketsApiticketGETApiPermissionTestMyTestStackTicketsApi42B25C87GETticketBB575261": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "GetTicket4BEE878E",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "TicketsApi5BB0C274",
              },
              "/test-invoke-stage/GET/ticket",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TicketsApiticketPOST4B2578D0": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "TicketsPost90C6DE87",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "TicketsApiticketA33D88A0",
        },
        "RestApiId": Object {
          "Ref": "TicketsApi5BB0C274",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "TicketsApiticketPOSTApiPermissionMyTestStackTicketsApi42B25C87POSTticket85FD9720": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "TicketsPost90C6DE87",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "TicketsApi5BB0C274",
              },
              "/",
              Object {
                "Ref": "TicketsApiDeploymentStageprod5BCA3CD4",
              },
              "/POST/ticket",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TicketsApiticketPOSTApiPermissionTestMyTestStackTicketsApi42B25C87POSTticket3EAE0373": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "TicketsPost90C6DE87",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "TicketsApi5BB0C274",
              },
              "/test-invoke-stage/POST/ticket",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TicketsPost90C6DE87": Object {
      "DependsOn": Array [
        "TicketsPostServiceRoleDefaultPolicy27D0FB29",
        "TicketsPostServiceRole8974BAF6",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersaefdea884930f8e31c77b692a0c9d29dc35671165502f5eb7303eae0df30f0f3S3BucketB7C90BB0",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersaefdea884930f8e31c77b692a0c9d29dc35671165502f5eb7303eae0df30f0f3S3VersionKey8F500E6E",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersaefdea884930f8e31c77b692a0c9d29dc35671165502f5eb7303eae0df30f0f3S3VersionKey8F500E6E",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "TABLE_NAME": Object {
              "Ref": "TicketsTableB76A19AF",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TicketsPostServiceRole8974BAF6",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TicketsPostServiceRole8974BAF6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TicketsPostServiceRoleDefaultPolicy27D0FB29": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "TicketsTableB76A19AF",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TicketsPostServiceRoleDefaultPolicy27D0FB29",
        "Roles": Array [
          Object {
            "Ref": "TicketsPostServiceRole8974BAF6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TicketsTableB76A19AF": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
        "TableName": "SttTickets",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
